### PR TITLE
Feat: cluster by

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: sqlmesh/core/snapshot/definition.py [hunk 1]
```diff
@@ -915,8 +915,11 @@ def _model_data_hash(model: Model) -> str:
         model.storage_format,
         str(model.lookback),
         *(model.partitioned_by or []),
+        *(model.clustered_by or []),
         model.stamp,
     ]
+    if model.clustered_by:
+        raise
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/definition.py:922]: Not sure I understand what we do here.

File: tests/core/test_model.py [hunk 2]
```diff
@@ -1231,23 +1233,23 @@ def test_star_expansion(assert_exp_eq) -> None:
...
     assert_exp_eq(
         context.models["db.model2"].render_query(snapshots=snapshots),
-        """
+        f"""
         SELECT
           "model1"."id" AS "id",
           "model1"."item_id" AS "item_id",
           "model1"."ds" AS "ds"
-        FROM "sqlmesh__db"."db__model1__3784582009" AS "model1"
+        FROM "sqlmesh__db"."db__model1__{snapshots['db.model1'].version}" AS "model1"
```
Comment (izeigerman) [inline tests/core/test_model.py:1241]: Did fingerprint change here because of your changes? Why? These models don't have clustered_by

**Comment by izeigerman** [comment]
Don't you need to update this: https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/model/definition.py#L1647
And also update model def validation logic here: https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/model/definition.py#L559 to bring it on par with partitioned_by.

File: tests/core/test_snapshot.py [hunk 3]
```diff
@@ -78,11 +78,12 @@ def test_json(snapshot: Snapshot):
     assert json.loads(snapshot.json()) == {
         "created_ts": 1663891973000,
         "ttl": "in 1 week",
-        "fingerprint": snapshot.fingerprint,
+        "fingerprint": snapshot.fingerprint.dict(),
```
Comment (izeigerman) [inline tests/core/test_snapshot.py:81]: wow, how did this work before?

File: tests/core/test_model.py (hunk 2 above)
Comment (tobymao) [reply tests/core/test_model.py:1241]: local sqlglot version triggered this, i decided to change this for future stuff

File: tests/core/test_snapshot.py (hunk 3 above)
Comment (tobymao) [reply tests/core/test_snapshot.py:81]: pydantic overrides equality checks

**Review by izeigerman — APPROVED** [review]

**Comment by crericha** [comment]
Snowflake supports `CLUSTER BY`. I think we should add that now or throw an exception if `clustered_by` is set for an engine that doesn't support it.

**Comment by crericha** [comment]
Looks like the spark dbt adapter uses a field called `clustered_by` instead of `cluster_by`. We'll need to support that field name too. I can do this after the PR is merged if you like.

**Comment by tobymao** [comment]
@crericha we'll need to add the bucket flag for spark with clustered_by

**Pull request merged**